File: core/c_functions/acquisition.c

#include <stdio.h>
#include <stdlib.h>
#include <complex.h>
#include "fft8g.h"

// GPS' definition of Pi
#define PI 3.1415926535898

#ifndef FFTMAX
#define FFTMAX 16384
#define FFTMAXSQRT 128
static int ip[FFTMAXSQRT + 2];   // needed for bit twiddling in FFT
static double w[FFTMAX * 5 / 4]; // needed for cos/sin weights in FFT
#endif

// --------------------------------------------------------------------------------------------------------------------

/*
@brief Pad an input signal to a power of two length and perform the FFT/IFFT on it.
@param signal Array with a real-valued signal (will be over-written with complex values).
@param size   Size of the array.
@param doInv  Whether to perform forward or inverse FFT.
@return void.
*/
void _fft(double* signal, const size_t size, const char doInv)
{
    // Compute the size to pad the signal to
    size_t paddedSize = 1;
    while (paddedSize < size)
        paddedSize <<= 1;

    // Pad the signal with zeroes
    double paddedSignal[paddedSize];
    for (size_t i = 0; i < size; i++)
        paddedSignal[i] = signal[i];
    for (size_t i = size; i < paddedSize; i++)
        paddedSignal[i] = signal[i];

    // Perform the FFT in place
    rdft(paddedSize, doInv == 0 ? 1 : -1, paddedSignal, ip, w);

    // Return the result in the same array
    for (size_t i = 0; i < size; i++)
        signal[i] = paddedSignal[i];

    return;
}

/*
@brief Pad an input signal to a power of two length and perform the real-valued FFT on it.
@param signal Array with a real-valued signal (will be over-written with complex values).
@param size   Size of the array.
@return void.
*/
void fft(double* signal, const size_t size)
{
    _fft(signal, size, 0);
    return;
}

/*
@brief Pad an input signal to a power of two length and perform the inverse FFT on it.
@param signal Array with a complex-valued signal (will be over-written with complex values).
@param size   Size of the array.
@return void.
*/
void ifft(double* signal, const size_t size)
{
    _fft(signal, size, 1);
    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief
@param
@return void.
@note Output array has size/2 complex entries.
*/
void setSatellite(const double* code,
                  const size_t size,
                  complex double* codeFFT)
{
    // Perform real-valued FFT on data in place
    double signal[size];
    for (size_t i=0; i < size; i++)
        signal[i] = code[i];
    fft(signal, size);

    // Conjugate and return the result in the output array
    for (size_t i=0; i < size; i++)
        ((double*) codeFFT)[i] = (i & 0x1) ? -signal[i] : signal[i];

    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief
@param
@return void.
@note Naive implementation returns the correlation map with half number of entries in the 
      second dimension than in the Python implementation due to the real-valued FFT. This 
      may affect the results.
*/
void PCPS(const complex double* rfData,
          const complex double* codeFFT, // 1D (samplesPerCode/2,)
          const long long cohIntegration,
          const long long nonCohIntegration,
          const long long samplesPerCode,
          const double samplingPeriod,
          const double interFrequency,
          const double* frequencyBins,
          const size_t s_frequencyBins,
          double* r_correlationMap) // 2D (s_frequencyBins, samplesPerCode/2)
{
    double phasePoints[cohIntegration * samplesPerCode];
    double signalCarrier[cohIntegration * samplesPerCode];
    complex double iqSignal[cohIntegration * samplesPerCode];
    double nonCohSum[samplesPerCode/2];
    complex double cohSum[samplesPerCode/2];
    for (size_t i=0; i < cohIntegration * samplesPerCode; i++)
        phasePoints[i] = (i << 1) * PI * samplingPeriod;
    
    // Do correlation for each frequency bin
    double freq;
    for (size_t i=0; i < s_frequencyBins; i++)
    {
        freq = interFrequency - frequencyBins[i];

        // Generate carrier replica
        for (size_t j=0; j < cohIntegration * samplesPerCode; j++)
            signalCarrier[j] = cexp(-I * freq * phasePoints[j]);

        // Non-coherent integration
        for (size_t j=0; j < samplesPerCode/2; j++)
            nonCohSum[j] = 0;
        for (size_t nonCohIdx=0; nonCohIdx < nonCohIntegration; nonCohIdx++)
        {
            // Mix the carrier with the required part of the data
            for (size_t j=0; j < cohIntegration * samplesPerCode; j++)
                iqSignal[j] = signalCarrier[j] * rfData[j + nonCohIdx*cohIntegration*samplesPerCode];

            // Coherent integration
            for (size_t j=0; j < samplesPerCode/3; j++)
                cohSum[j] = 0;
            for (size_t cohIdx=0; cohIdx < cohIntegration; cohIdx++)
            {
                // Perform FFT (in-place)
                fft((double*)(iqSignal + cohIdx * samplesPerCode), samplesPerCode);

                // Correlate with C/A code (in-place)
                for (size_t j=0; j < samplesPerCode/2; j++)
                    ((complex double*)(iqSignal + cohIdx * samplesPerCode))[j] *= codeFFT[j];

                // Perform IFFT
                ifft((double*)(iqSignal + cohIdx * samplesPerCode), samplesPerCode);
                for (size_t j=0; j < samplesPerCode/2; j++)
                    cohSum[j] = ((complex double*)(iqSignal + cohIdx * samplesPerCode))[j];
            }
            for (size_t j=0; j < samplesPerCode/2; j++)
                nonCohSum[j] += cabs(cohSum[j]);
        }
        for (size_t j=0; j < samplesPerCode/2; j++)
            r_correlationMap[i * s_frequencyBins + j] = nonCohSum[j];
    }

    return;
}

// --------------------------------------------------------------------------------------------------------------------

/*
@brief
@param
@return void.
*/
void twoCorrelationPeakComparison(const double* correlationMap, // 2D (s_frequencyBins, s_correlationMap)
                                  const size_t s_correlationMap,
                                  const double* frequencyBins,
                                  const size_t s_frequencyBins,
                                  const long long samplesPerCode,
                                  const long long samplesPerCodeChip,
                                  const double interFrequency,
                                  double* r_acquisitionMetric,
                                  double* r_estimatedDoppler,
                                  double* r_estimatedFrequency,
                                  long long* r_estimatedCode,
                                  long long* r_idxEstimatedFrequency,
                                  long long* r_idxEstimatedCode)
{
    double peak1 = 0.0, peak2 = 0.0;
    size_t idx0, idx1;
    long long exclude0, exclude1;

    // Find first correlation peak
    for (size_t i=0; i < s_frequencyBins * s_correlationMap; i++)
    {
        if (correlationMap[i] > peak1)
        {
            peak1 = correlationMap[i];
            idx0 = i / s_correlationMap;
            idx1 = i % s_correlationMap;
        }
    }

    // Find second correlation peak
    exclude0 = idx1 - samplesPerCodeChip;
    exclude1 = idx1 + samplesPerCodeChip;
    if (exclude0 < 1)
    {
        for (size_t i=idx0 * s_correlationMap + exclude1; i < (idx0 + 1) * s_correlationMap; i++)
            if (correlationMap[i] > peak2)
                peak2 = correlationMap[i];
    }
    else if (exclude1 >= samplesPerCode)
    {
        for (size_t i=idx0 * s_correlationMap; i < idx0 * s_correlationMap + exclude0; i++)
            if (correlationMap[i] > peak2)
                peak2 = correlationMap[i];
    }
    else
    {
        for (size_t i=idx0 * s_correlationMap; i < idx0 * s_correlationMap + exclude0; i++)
            if (correlationMap[i] > peak2)
                peak2 = correlationMap[i];
        for (size_t i=idx0 * s_correlationMap + exclude1; i < (idx0 + 1) * s_correlationMap; i++)
            if (correlationMap[i] > peak2)
                peak2 = correlationMap[i];
    }

    // Compute acquisition metric
    *r_estimatedDoppler = -frequencyBins[idx0];
    *r_estimatedCode = idx1;
    *r_acquisitionMetric = peak1 / peak2;
    *r_idxEstimatedFrequency = idx0;
    *r_estimatedFrequency = interFrequency + *r_estimatedDoppler;
    *r_idxEstimatedCode = idx1;

    return;
}

/// ===================================================================================================================

int main()
{
    printf("Hello World!\n");

    return 0;
}

File: core/dsp/acquisition.py

import numpy as np

# =====================================================================================================================

def PCPS(rfData:np.array, interFrequency:float, samplingFrequency:float, codeFFT:np.array, dopplerRange:tuple, 
         dopplerStep:int, samplesPerCode:int, coherentIntegration:int=1, nonCoherentIntegration:int=1):
    """
    Implementation of the Parallel Code Phase Search (PCPS) method [Borre, 2007]. This method perform the correlation 
    of the code in the frequency domain using FFTs. It produces a 2D correlation map over the frequency and code 
    dimensions.

    Args:
        rfData (numpy.array): Data sample to be used.
        interFrequency (float) : Intermediate Frequency used in RF signal.
        codeFFT (numpy.array): Primary code of the GNSS signal, transformed using a FFT. 
        dopplerRange (int): Frequency bound (-/+) for acquisition search.
        dopplerStep (int) : Frequency step for acquisition search.
        codeBins (np.array): Code bins 

    Returns:
        correlationMap (numpy.array): 2D correlation results.

    Raises:
        None
    """

    rfData = np.squeeze(rfData)

    phasePoints = np.array(range(coherentIntegration * samplesPerCode)) * 2 * np.pi / samplingFrequency
    frequencyBins = np.arange(-dopplerRange, dopplerRange, dopplerStep)

    # Search loop
    correlationMap = np.zeros((len(frequencyBins), samplesPerCode))
    noncoh_sum     = np.zeros((1, samplesPerCode))
    coh_sum        = np.zeros((1, samplesPerCode))
    idx = 0
    for freq in frequencyBins:
        freq = interFrequency - freq

        # Generate carrier replica
        signal_carrier = np.exp(-1j * freq * phasePoints)

        # Non-Coherent Integration 
        noncoh_sum = noncoh_sum * 0.0
        for idx_noncoh in range(0, nonCoherentIntegration):
            # Select only require part of the dataset
            iq_signal = rfData[idx_noncoh*coherentIntegration*samplesPerCode:(idx_noncoh+1)*coherentIntegration*samplesPerCode]
            # Mix with carrier
            iq_signal = np.multiply(signal_carrier, iq_signal)
            
            # Coherent Integration
            coh_sum = noncoh_sum * 0.0
            for idx_coh in range(0, coherentIntegration):
                # Perform FFT
                iq_fft = np.fft.fft(iq_signal[idx_coh*samplesPerCode:(idx_coh+1)*samplesPerCode])

                # Correlation with C/A code
                iq_conv = np.multiply(iq_fft, codeFFT)

                # Inverse FFT (go back to time domain)
                coh_sum = coh_sum + np.fft.ifft(iq_conv)

            # Absolute values
            noncoh_sum = noncoh_sum + abs(coh_sum)
        
        correlationMap[idx, :] = abs(noncoh_sum)
        idx += 1
    correlationMap = np.squeeze(np.squeeze(correlationMap))

    return correlationMap

# =====================================================================================================================

def TwoCorrelationPeakComparison(correlationMap:np.array, samplesPerCode:int, samplesPerCodeChip:int):
    """ 
    Perform analysis on correlation map, finding the the highest peak and comparing its correlation value to the one 
    from the second highest peak.

    Args:
        correlationMap (numpy.array): 2D-array from correlation method.
        samplesPerCode (int): Number of samples per code.
        samplesPerCodeChip (int): Number of code samples per code chip
    
    Returns:
        idxHighestPeak (tuple): Indices of the highest correlation peak. 
        acquisitionMetric (float): Ratio between the highest and second highest peaks.
    
    Raises:
        None

    """
    
    # Find first correlation peak
    peak_1 = np.amax(correlationMap)
    idx = np.where(correlationMap == peak_1)
    idxHighestPeak = (int(idx[0]), int(idx[1]))

    # Find second correlation peak
    exclude = list((int(idxHighestPeak[1] - samplesPerCodeChip), int(idxHighestPeak[1] + samplesPerCodeChip)))

    if exclude[0] < 1:
        code_range = list(range(exclude[1], samplesPerCode - 1))
    elif exclude[1] >= samplesPerCode:
        code_range = list(range(0, exclude[0]))
    else:
        code_range = list(range(0, exclude[0])) + list(range(exclude[1], samplesPerCode - 1))
    peak_2 = np.amax(correlationMap[idxHighestPeak[0], code_range])
    
    acquisitionMetric = peak_1 / peak_2

    return idxHighestPeak, acquisitionMetric

# =====================================================================================================================
